# Importing the libraries

In [ ]:
import pandas as pd
import numpy as np

# Reading the dataset

In [ ]:
ratings = pd.read_csv("/content/jokes-data.csv", sep = ",")

In [ ]:
ratings.head()

id  user_id  joke_id  Rating
0  31030_110    31030      110   2.750
1  16144_109    16144      109   5.094
2    23098_6    23098        6  -6.438
3   14273_86    14273       86   4.406
4  18419_134    18419      134   9.375

In [ ]:
ratings.shape

(189591, 4)

In [ ]:
ratings['user_id'].nunique()

40863

In [ ]:
ratings['joke_id'].nunique()

139

In [ ]:
ratings['id'].nunique()

189591

# Checking for null values

In [ ]:
ratings.isna().sum()

id         0
user_id    0
joke_id    0
Rating     0
dtype: int64

# Installing 'Surprise'

In [ ]:
##### surprise

In [ ]:
!pip install surprise

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#### collobarative filteing

In [ ]:
#### user_id, joke_id, rating

# Dropping the column 'id'

In [ ]:
ratings = ratings.drop("id", axis=1)

In [ ]:
ratings

user_id  joke_id  Rating
0       31030.0    110.0   2.750
1       16144.0    109.0   5.094
2       23098.0      6.0  -6.438
3       14273.0     86.0   4.406
4       18419.0    134.0   9.375
...         ...      ...     ...
189586  16952.0    100.0   2.656
189587  12484.0     40.0   8.844
189588  30281.0    134.0  -2.281
189589  23364.0    107.0   1.062
189590      NaN      NaN     NaN

[189591 rows x 3 columns]

In [ ]:
from surprise import Reader, Dataset, KNNBasic, SVD

In [ ]:
##### define a reader object

In [ ]:
reader = Reader()

In [ ]:
#### create a the dataset to be used for building the filter

In [ ]:
data = Dataset.load_from_df(ratings, reader)

In [ ]:
data

In [ ]:
#algo = KNNBasic()

In [ ]:
from surprise.model_selection import cross_validate

In [ ]:
#cross_validate(algo, data, measures = ["RMSE"], cv = 5, verbose = True)

## My system crashed several times when I tried cross validation on KNNBasic.

# Cross Validation on SVD

In [ ]:
algo2 = SVD()

In [ ]:
cross_validate(algo2, data, measures = ["RMSE"], cv = 5, verbose = True)

Evaluating RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    4.6503  4.6568  4.6597  4.6701  4.6723  4.6618  0.0082  
Fit time          20.59   21.64   21.95   23.31   21.23   21.74   0.90    
Test time         3.81    3.35    2.96    3.09    3.55    3.35    0.31    


{'test_rmse': array([4.65032058, 4.65680611, 4.65968409, 4.67012007, 4.67226846]),
 'fit_time': (20.594974517822266,
  21.6440749168396,
  21.946969509124756,
  23.306283712387085,
  21.226240396499634),
 'test_time': (3.8135039806365967,
  3.3476147651672363,
  2.9601190090179443,
  3.0948736667633057,
  3.546802043914795)}

# Conclusion:
## SVD is rating the unknown rating of the sparse matrix with mean RMSE as 4.66.